src/electron/preload.ts

```diff
@@ -0,0 +1,81 @@
+// Copyright 2018 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Starting from electron 12, contextIsolation is turned on by default.
+// So we are not able to use node APIs or electron APIs in renderer scripts
+// any more. We have to inject key features into the global window object
+// in preload: https://www.electronjs.org/docs/latest/tutorial/tutorial-preload.
+
+import {contextBridge, ipcRenderer} from 'electron';
+import * as os from 'os';
+import {ShadowsocksConfig} from '../www/app/config';
+import {ErrorCode} from '../www/model/errors';
+
+contextBridge.exposeInMainWorld('os', {
+  platform: os.platform(),
+});
+
+contextBridge.exposeInMainWorld('ipc', {
```

Comment:
I found a library doing some type checks between IPC handlers and callers: https://github.com/JichouP/electron-typescript-ipc. But the code does not seem to be intuitive.

Reply:
I wasn't proposing anything so complicated, just something like:

```ts
const ipc: IIpcAPI = {
...
};

contextBridge.exposeInMainWorld('ipc', ipc);
```

This would require moving the interface definition to a file that can be referenced in both contexts, but otherwise it should be trivial, and would help to ensure that changes to the API are consistent across both contexts.

Reply:
Yes, now all IPC messages are strongly typed which are defined in `ipc.ts`. And developers only need to update this file to add/update/remove IPC messages (which means, they don't need to change other files like `preload.ts`). Now the TypeScript compiler will show any incompatible type usages after the change so the developers can easily fix them.